### ISSUE TITLE
Failure to guess when a generic function is used as an argument

### ISSUE THREAD

**Issue by yassun7010** [microsoft/pyright#6326]
**Describe the bug**
I have found cases where type inference does not work well for functions that take generic functions as arguments.

**Code or Screenshots**

case 1 : Generic Class

```python
from typing import Callable

class A[T]:
    def __init__(self, value: T) -> None:
        self.value = value

def do_something[T](callable: Callable[[type[A[T]]], A[T]]) -> T:
    return callable(A).value

a: A[int] = A(12) # pass

a1: int = do_something(lambda A: A(12)) # pass
a2 = do_something(lambda A: A(12)) # failed: Type "Literal[12]" cannot be assigned to type "T@do_something" (reportGeneralTypeIssues)
```

case 2: Generic Callable

```python
from typing import Callable

def do_something2[T](callable: Callable[[Callable[[T], T]], T]) -> T:
    return callable(lambda x: x)

a1: int = do_something2(lambda a: a(12)) # pass
a2 = do_something2(lambda a: a(12)) # failed: Type "Literal[12]" cannot be assigned to type "T@do_something2" (reportGeneralTypeIssues)
```

**VS Code extension or command-line**

```txt
Using python3 (3.12.0)
pyright 1.1.334
```

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.

This is a difficult case for the type inference logic to handle. FWIW, mypy doesn't work with this code sample either.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.337, which I just published. It will also be included in a future release of pylance.
